### DISCUSSION TITLE
IJob Constructor take in Dependency Injected parameters

### DISCUSSION THREAD

**Discussion by travis-pettry1** [quartznet/quartznet#1651] [Q&A]
Problem:
I cannot inject dependencies into my Ijob's constructor. 

Solution:
I would like to be able to inject dependencies into the constructor of my jobs. This would allow me to use services that I have defined in my jobs.

Alternatives:
I have not been able to find an alternative.

**Comment by lahma** [comment]
Have you checked the examples, like [the ASP.NET Core sample](https://github.com/quartznet/quartznet/tree/main/src/Quartz.Examples.AspNetCore)?

**Reply by travis-pettry1** [reply]
@lahma  I have tried that, however, when I have a job that does not have a parameterless constructor my jobs do not get executed.

**Reply by lahma** [reply]
You should enable logging and see what is happening.  Also compare carefully the differences between your configuration and the example app.

**Reply by travis-pettry1** [reply]
@lahma  This is from the logging.

[08:03:29 ERR] An error occurred instantiating job to be executed. job= 'DEFAULT.TestPlugin1'
Quartz.SchedulerException: Problem instantiating class 'SDS2.JobScheduler.PluginTest.TestPlugin1: Cannot instantiate type which has no empty constructor (Parameter 'TestPlugin1')'
 ---> System.ArgumentException: Cannot instantiate type which has no empty constructor (Parameter 'TestPlugin1')
   at Quartz.ExceptionHelper.ThrowArgumentException(String message, String paramName)
   at Quartz.Util.ObjectUtils.InstantiateType[T](Type type)
   at Quartz.Simpl.SimpleJobFactory.NewJob(TriggerFiredBundle bundle, IScheduler scheduler)
   --- End of inner exception stack trace ---
   at Quartz.Simpl.SimpleJobFactory.NewJob(TriggerFiredBundle bundle, IScheduler scheduler)
   at Quartz.Simpl.PropertySettingJobFactory.InstantiateJob(TriggerFiredBundle bundle, IScheduler scheduler)
   at Quartz.Simpl.PropertySettingJobFactory.NewJob(TriggerFiredBundle bundle, IScheduler scheduler)
   at Quartz.Core.JobRunShell.Initialize(QuartzScheduler sched, CancellationToken cancellationToken) [See nested exception: System.ArgumentException: Cannot instantiate type which has no empty constructor (Parameter 'TestPlugin1')
   at Quartz.ExceptionHelper.ThrowArgumentException(String message, String paramName)
   at Quartz.Util.ObjectUtils.InstantiateType[T](Type type)
   at Quartz.Simpl.SimpleJobFactory.NewJob(TriggerFiredBundle bundle, IScheduler scheduler)]
   
   
   My Constructor 
   
   public TestPlugin1(ILogger<TestPlugin1> logger)
        {
            _logger = logger;   
        }

**Reply by lahma** [reply]
Your configuration is wrong. MS DI support isn't activated, it's using the default job factory.

**Reply by travis-pettry1** [reply]
@lahma I am using Quartz as a hosted service what would be the best way to enable that 

```
var props = new NameValueCollection
{
   {"quartz.jobStore.type", "Quartz.Impl.AdoJobStore.JobStoreTX, Quartz" },
   {"quartz.jobStore.lockHandler.type", "Quartz.Impl.AdoJobStore.UpdateLockRowSemaphore, Quartz"},
   {"quartz.jobStore.useProperties", "true"},
   {"quartz.jobStore.dataSource", "default"},
   {"quartz.jobStore.tablePrefix", "QRTZ_"},
   {"quartz.jobStore.driverDelegateType", "Quartz.Impl.AdoJobStore.SQLiteDelegate, Quartz"},
   {"quartz.dataSource.default.provider", "SQLite"},
   {"quartz.dataSource.default.connectionString", "Data Source=myDB.db;Version=3;Foreign Keys=ON;"},
   {"quartz.serializer.type", "json"},
   {"quartz.scheduler.typeLoadHelper.type" , typeof(MyTypeLoader).AssemblyQualifiedName },
   {"quartz.plugin.recentHistory.type", "Quartz.Plugins.RecentHistory.ExecutionHistoryPlugin, Quartz.Plugins.RecentHistory" },
   {"quartz.plugin.recentHistory.storeType", "Quartz.Plugins.RecentHistory.Impl.InProcExecutionHistoryStore, Quartz.Plugins.RecentHistory" }
};

 services.AddSingleton<ISchedulerFactory>(new StdSchedulerFactory(props));
// Quartz.Extensions.Hosting hosting
services.AddQuartzHostedService(options =>
{
   // when shutting down we want jobs to complete gracefully
   options.WaitForJobsToComplete = true;
});    
```

**Reply by lahma** [reply]
So you didn't compare the configuration example I gave you?

**Reply by travis-pettry1** [reply]
@lahma  I did. In the provided example it is using addQuartz. This method gives you access to UseMicrosoftDependencyInjectionJobFactory(); I do not have access to that method call due to the fact I am using quartz as a hosted service.

**Reply by travis-pettry1** [reply]
@lahma Also, the MicrosoftDependencyInjectionJobFactory doesn't appear to allow the custom config that I have listed above

**Reply by lahma** [reply]
Maybe you need to check the worker example then. It's using the hosted service.

https://github.com/quartznet/quartznet/blob/98653375cdc1861b4b98719ce6471c7e632f2490/src/Quartz.Examples.Worker/Program.cs#L86-L94

**Reply by travis-pettry1** [reply]
@lahma thanks for your help, I was able to get it worked out.
